internal/aws/ssm_remote_access_strategy.go

```diff
@@ -146,64 +245,144 @@ func (s *SSMRemoteAccessStrategy) CleanupSSMManagedNodes(ctx context.Context, po
+	logger.V(1).Info("Creating SSM activation")
 	activationCode, activationId, err := s.createSSMActivation(ctx, pod, workspace, accessStrategy)
 	if err != nil {
 		return fmt.Errorf("failed to create SSM activation: %w", err)
 	}
+	logger.Info("SSM activation created", "activationId", activationId)
 
-	// Step 2: Run register-ssm.sh with sensitive values passed via stdin
-	logger.Info("Running SSM registration script in sidecar")
+	// Run registration script
+	logger.V(1).Info("Running SSM registration script in sidecar")
 	region := s.ssmClient.GetRegion()
-
 	// Use stdin to pass only sensitive values securely
 	cmd := []string{"bash", "-c", fmt.Sprintf("read ACTIVATION_ID && read ACTIVATION_CODE && env ACTIVATION_ID=\"$ACTIVATION_ID\" ACTIVATION_CODE=\"$ACTIVATION_CODE\" REGION=%s %s", region, SSMRegistrationScript)}
 	stdinData := fmt.Sprintf("%s\n%s\n", activationId, activationCode)
 
 	if _, err := s.podExecUtil.ExecInPod(ctx, pod, SSMAgentSidecarContainerName, cmd, stdinData); err != nil {
 		return fmt.Errorf("failed to execute SSM registration script: %w", err)
 	}
+	logger.Info("SSM registration completed successfully")
 
-	// Step 3: Start remote access server in main container
-	logger.Info("Starting remote access server in main container")
-	serverCmd := []string{"bash", "-c", fmt.Sprintf("sudo %s > /dev/null 2>&1 &", RemoteAccessServerPath)}
-	if _, err := s.podExecUtil.ExecInPod(ctx, pod, WorkspaceContainerName, serverCmd, noStdin); err != nil {
-		return fmt.Errorf("failed to start remote access server: %w", err)
-	}
-
-	// Step 4: Create completion marker file
-	logger.Info("Creating SSM registration completion marker")
+	// TODO: Remove this once all deployments migrate to state-file-based readiness checks
+	logger.V(1).Info("Creating marker file for backward compatibility")
 	markerCmd := []string{"touch", SSMRegistrationMarkerFile}
 	if _, err := s.podExecUtil.ExecInPod(ctx, pod, SSMAgentSidecarContainerName, markerCmd, noStdin); err != nil {
-		return fmt.Errorf("failed to create completion marker: %w", err)
+		return fmt.Errorf("failed to create marker file: %w", err)
+	}
+
+	logger.Info("Sidecar container setup completed")
+	return nil
+}
+
+// setupWorkspaceContainer starts the remote access server
+func (s *SSMRemoteAccessStrategy) setupWorkspaceContainer(ctx context.Context, pod *corev1.Pod) error {
+	logger := logf.FromContext(ctx).WithValues("pod", pod.Name)
+
+	// Check workspace is running
+	if !s.isContainerRunning(pod, WorkspaceContainerName) {
+		logger.V(1).Info("Workspace container not running yet, waiting")
+		return nil
+	}
+
+	// TODO: Make this script idempotent
+	// For now, only called on container restart so should be fine
+	logger.V(1).Info("Starting remote access server in workspace container")
+	serverCmd := []string{"bash", "-c", fmt.Sprintf("sudo %s > /dev/null 2>&1 &", RemoteAccessServerPath)}
+	if _, err := s.podExecUtil.ExecInPod(ctx, pod, WorkspaceContainerName, serverCmd, ""); err != nil {
```

Comment:
Not blocking. 
We can ignore the error when the port is already alive instead of print error.